### PR TITLE
Fixing cleanUp method to restore environment

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

The clean-up method was not reliabliy restoring sys.environment on
windows. Using Mock's patch to cleanup more reliably.

### What issues does this PR fix or reference?

#46350, #46345, #46349, #46354

### Previous Behavior

cleanUp method was not restoring sys.environ properly (at least on windows)

### New Behavior

Mock's patch method restores environ properly

### Tests written?

No - This fixes a hand full of existing tests.

### Commits signed with GPG?

Yes

### CODE REVIEW

**Comment by dwoz** [comment]
Looks like this is breaking some other tests in unit.states.test_environ. Making this a WIP until I resolve those.

**Comment by dwoz** [comment]
Adding `clean=True` to the patch call seems to have done the trick. I think this is good to go.

**Comment by dwoz** [comment]
I ended up modifying `states.environ.setenv` to handle windows environments a bit more reasonably. The crux of the issue is that environment variables on windows are not case sensitive. Keys are always represented as upper case however.

Without my patch the `tests.unit.states.test_environ.TestEnvironState.test_setenv` test will always fail on the third assertion because `setenv` was in-correctly updating the key `test` which already exists in the environment.

**Review by twangboy — APPROVED** [review]

**Comment by rallytime** [comment]
@dwoz Looks like there's a related test failure here and some lint. Can you take another look?

**Comment by rallytime** [comment]
@dwoz This is still failing a related test on linux: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/21135/

**Comment by dwoz** [comment]
I'm hoping we'll be good after the last commit.

**Comment by dwoz** [comment]
re-run all

**Comment by rallytime** [comment]
@dwoz One more lint fix :)

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/20489/violations/file/tests/unit/states/test_environ.py/

**Comment by dwoz** [comment]
@rallytime That linter is doing it's job.

**Comment by rallytime** [comment]
re-run py

File: tests/unit/states/test_environ.py [hunk 1]
```diff
@@ -6,14 +6,16 @@
...
+from tests.support.unit import TestCase, skipIf
 from tests.support.mock import (
     MagicMock,
     patch
 )
 # Import salt libs
 import salt.states.environ as envstate
 import salt.modules.environ as envmodule
+import salt.modules.reg as regmodule
+import salt.utils as utils
```
Comment (rallytime) [inline tests/unit/states/test_environ.py]: I know the `environ` state above was doing it this way, but we are trying to make these imports more uniform. Can you change this to use the full path? That makes it easier to lint/check for these references in newer branches that have diverged from using the funcs in the `salt.utils.__init__.py` file.

File: tests/unit/states/test_environ.py [hunk 2]
```diff
@@ -47,9 +53,10 @@ def test_setenv(self):
         ret = envstate.setenv('test', 'other')
         self.assertEqual(ret['changes'], {})
 
+    @skipIf(not utils.is_windows(), 'Windows only')
     def test_setenv_permanent(self):
-        with patch.dict(envmodule.__salt__, {'reg.set_value': MagicMock(), 'reg.delete_value': MagicMock()}), \
-                patch('salt.utils.is_windows', MagicMock(return_value=True)):
+        '''test that we can set perminent environment variables (requires pywin32)'''
```
Comment (rallytime) [inline tests/unit/states/test_environ.py]: Can you split up this doc line into 3 lines?

**Review by rallytime — CHANGES_REQUESTED** [review]
Just some stylist things that need to be changed, otherwise 👍

**Comment by rallytime** [comment]
re-run py2

**Review by rallytime — APPROVED** [review]

**Pull request merged**